src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1500,46 +1497,49 @@ private IEnumerable<TypeInferenceInfo> GetPatternTypes(PatternSyntax pattern)
                 {
                     case ConstantPatternSyntax constantPattern: return GetTypes(constantPattern.Expression);
                     case DeclarationPatternSyntax declarationPattern: return GetTypes(declarationPattern.Type);
-                    case DeconstructionPatternSyntax deconstructionPattern: return GetTypesForDeconstructionPattern(deconstructionPattern);
-                    case PropertyPatternSyntax propertyPattern: return GetTypes(propertyPattern.Type);
+                    case RecursivePatternSyntax recursivePattern: return GetTypesForRecursivePattern(recursivePattern);
                     default: return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
                 }
             }
 
-            private IEnumerable<TypeInferenceInfo> GetTypesForDeconstructionPattern(DeconstructionPatternSyntax deconstructionPattern)
+            private IEnumerable<TypeInferenceInfo> GetTypesForRecursivePattern(RecursivePatternSyntax recursivePattern)
             {
                 // if it's of the for "X (...)" then just infer 'X' as the type.
-                if (deconstructionPattern.Type != null)
+                if (recursivePattern.Type != null)
                 {
-                    return CreateResult(SemanticModel.GetTypeInfo(deconstructionPattern).ConvertedType);
+                    return CreateResult(SemanticModel.GetTypeInfo(recursivePattern).ConvertedType);
                 }
 
                 // If it's of the form (...) then infer that the type should be a 
                 // tuple, whose elements are inferred from the individual patterns
                 // in the deconstruction.
-                var subPatternCount = deconstructionPattern.SubPatterns.Count;
-                if (subPatternCount >= 2)
+                var deconstructionPart = recursivePattern.DeconstructionSubpattern;
+                if (deconstructionPart != null)
                 {
-                    // infer a tuple type for this deconstruction.
-                    var elementTypesBuilder = ArrayBuilder<ITypeSymbol>.GetInstance(subPatternCount);
-                    var elementNamesBuilder = ArrayBuilder<string>.GetInstance(subPatternCount);
-
-                    foreach (var subPattern in deconstructionPattern.SubPatterns)
+                    var subPatternCount = deconstructionPart.SubPatterns.Count;
+                    if (subPatternCount >= 2)
```

Comment:
could simplify as ```if (deconstructionPart ?.SubPatterns.Count >= 2)```

Reply:
Yes, but the count variable is reused.

---
In reply to: [187493907](https://github.com/dotnet/roslyn/pull/26782#discussion_r187493907) [](ancestors = 187493907)